analysis.go

```diff
@@ -50,60 +49,51 @@ type DeploymentConfigs struct {
 	} `json:"items"`
 }
 
-type CheckTask func(string, io.Writer) (Result, error)
+type CheckTask func(string, string) (Result, error)
+
+func GetAnalysisTasks(tasks chan<- Task, basepath string, projects []string, results chan<- CheckResults) {
+	// Platform-wide analysis goes here
+
+	// project specific analysis in here
+	for _, p := range projects {
+		tasks <- CheckProjectTask(p, basepath, results)
+	}
+}
 
 // CheckTasks is a task factory for tasks that diagnose system conditions.
-func CheckTasks(project string, outFor, errOutFor projectResourceWriterCloserFactory) Task {
-	return checkTasks(func() []CheckTask {
+func CheckProjectTask(project, basepath string, results chan<- CheckResults) Task {
+	return checkProjectTask(func() []CheckTask {
 		return []CheckTask{CheckImagePullBackOff, CheckDeployConfigsReplicasNotZero}
-	}, project, outFor, errOutFor)
+	}, project, basepath, results)
 }
 
 // A getProjectCheckFactory generates tasks to diagnose system conditions.
 type getProjectCheckFactory func() []CheckTask
 
 type CheckResults struct {
+	Scope   string `json:"scope"`
 	Results []Result
 }
 
 // checkTasks executes all the CheckTasks returned from the supplied
 // checkFactory against the specified project. The results of the checks are
-// combined into a single JSON object and written to the writer returned from
-// outFor, any errors that occur during the test are written to the writer
-// returned from errOutFor.
-func checkTasks(checkFactory getProjectCheckFactory, project string, outFor, errOutFor projectResourceWriterCloserFactory) Task {
+// combined into a single JSON object and returned
+func checkProjectTask(checkFactory getProjectCheckFactory, project, basepath string, results chan<- CheckResults) Task {
 	return func() error {
-		stdOut, stdOutCloser, err := outFor(project, "analysis")
-		if err != nil {
-			return err
-		}
-		defer stdOutCloser.Close()
-		stdErr, stdErrCloser, err := errOutFor(project, "analysis")
-		if err != nil {
-			return err
-		}
-		defer stdErrCloser.Close()
-
-		results := CheckResults{Results: []Result{}}
-
```

Comment:
This code felt much better by the time this PR was finished. Can't wait to improve it even more!